CB-8761 [wp8]: Copy cookies from WebBrowser

dpolivy opened the pull request:
On other platforms, the cookies from the browser that apply to the
request are automatically included. On WP8, that's not the case.

This commit will copy the cookies from the WebBrowser control, if the
scheme and host of the WebBrowser match that of the HttpWebRequest. For
users who host their web components remotely, and perform
uploads/downloads from the same server using cookie based
authentication, this fix enables that scenario on WP8, and brings it up
to parity with iOS, Android, and others.

Fixes https://issues.apache.org/jira/browse/CB-8761

dpolivy commented:
@stevengill Is there someone who can review and merge this?

nikhilkh commented:
@vladimir-kotikov Can you please take a look?

vladimir-kotikov commented on src/wp/FileTransfer.cs:
```diff
@@ -210,6 +216,77 @@ public FileTransferProgress(long bTotal = 0, long bLoaded = 0)
+                    // Only copy the cookies if the scheme and host match (to avoid any issues with secure/insecure cookies)
+                    // NOTE: since the returned CookieCollection appears to munge the original cookie's domain value in favor of the actual Source domain,
+                    // we can't know for sure whether the cookies would be applicable to any other hosts, so best to play it safe and skip for now.
+                    if (this.browser.Source.Scheme == webRequest.RequestUri.Scheme && this.browser.Source.Host == webRequest.RequestUri.Host)
```
it is very unlikely, but `browser` still can be null here, consider adding check for 'not null' case.

Also `browser.Source.Scheme` and `browser.Source.Host` will throw an `InvalidOperationException` if `Source` is an relative URI. IMO it's better to check if `browser.Source.IsAbsoluteUri == true` than catch an exception here.

vladimir-kotikov commented on src/wp/FileTransfer.cs:
```diff
@@ -210,6 +216,77 @@ public FileTransferProgress(long bTotal = 0, long bLoaded = 0)
+                        {
+                            // Check that the path is allowed, first
+                            // NOTE: Path always seems to be empty for now, even if the cookie has a path set by the server.
+                            if (cookie.Path.Length == 0 || requestPath.IndexOf(cookie.Path) == 0)
```
`requestPath.IndexOf(cookie.Path)` - this comparison is case-sensitive. Consider using overload of `indexOf` method with `InvariantCultureIgnoreCase` or `OrdinalIgnoreCase` comparison.

vladimir-kotikov commented on src/wp/FileTransfer.cs:
```diff
@@ -210,6 +216,77 @@ public FileTransferProgress(long bTotal = 0, long bLoaded = 0)
+                }
+                catch (Exception ex)
+                {
+                    tcs.SetException(ex);
```
This exception will never be caught. Consider silently return from this method here or add try/catch block around this method call in `upload` and `download`.

dpolivy commented:
@vladimir-kotikov Thanks for the comments! I addressed your feedback, and pushed up a new diff. I squashed the changes into a single commit -- hope that is OK (I'm not sure what the correct process is for that, but hopefully this makes it an easier merge for you).

sgrebnov commented:
LGTM

sgrebnov commented on src/wp/FileTransfer.cs:
```diff
@@ -210,6 +216,80 @@ public FileTransferProgress(long bTotal = 0, long bLoaded = 0)
+            var tcs = new TaskCompletionSource<object>();
+
+            // Accessing WebBrowser needs to happen on the UI thread
+            Deployment.Current.Dispatcher.BeginInvoke(() =>
```
I think we can use Invoke() instead of BeginInvoke() here so we don't need to additionally use TaskCompletionSource.

dpolivy replied:
@sgrebnov I think the Dispatcher only has the async methods on WP, unfortunately. If you have another suggestion, though, I'm happy to look at it.

sgrebnov replied:
@dpolivy I've double checked and you are right, Invoke does not exit in WP8 .Net framework subset.  So please ignore my comment..
